### DISCUSSION TITLE
RPC FastAPI

### DISCUSSION THREAD

**Discussion by Dester507** [fastapi/fastapi#6974] [Questions]
Can i call APIRoute func if i have only function name and params?

I have middleware that accept requests, with rpc-xml data. After that i parse this xml and get from it func name and params.

```python
@rpc_api.middleware("http")
async def process_time_handler(request: Request, call_next):
    if request.url == "http://127.0.0.1:8000/xml/":
        try:
            xml_str = await request.body()
            response = await handle_xml(xml_str)
        except NameError:
            return JSONResponse(status_code=404, content="Method does not exist")
        except TypeError:
            return JSONResponse(status_code=404, content="Error with arguments")
        except SyntaxError:
            return JSONResponse(status_code=404, content="Method name is void")
        try:
            if isinstance(response, dict):
                response = JSONResponse(response)
            return response
        except:
            return JSONResponse(status_code=404, content="Bad response")
    else:
        response = await call_next(request)
        return response
```
How can i call APIRoute func?

Now i call func by getattr(module, func-name)(*args, **kwargs), but i want call by route.

```python
async def handle_xml(xml_text):
    try:
        from type_plugin import PathOperationFactory
        xml_request = await parser_run(xml_text)
        method_name = xml_request.xpath("//methodName[1]")[0].text
        kwargs = {}
        args = list(
            map(
                xml2py,
                xml_request.xpath("//params/param/value")
            )
        )
        if isinstance(args[-1], dict):
            kwargs = args.pop(-1)
        method_name = method_name.split('_', 1)
        method_namespace = f"routes.{method_name[0]}"
        response = await PathOperationFactory.factory(method_namespace, method_name[1], *args, **kwargs)
        if response is None:
            return await create_bad_response("Method does not exist.")
        return await create_good_response(response)
    except TypeError as ex:
        print(ex)
        return await create_bad_response("Error with arguments")
    except Exception as ex:
        print(ex, type(ex))
        return await create_bad_response("Error with parse xml request.")
```

```python
class PathOperationFactory:
    @staticmethod
    async def factory(namespace: str, method: str, *args, **kwargs) -> Optional[Callable]:
        try:
            return await getattr(PLUGINS[namespace], method)(*args, **kwargs)
        except (KeyError, AttributeError):
            return None
```

**Comment by falkben** [comment]
Don't you need to pass the request on through? What if you mult. middlware? `call_next(request)` probably what you want. You should also be aware that processing the body in the middleware is something to try and avoid.  Instead, create a custom request: https://fastapi.tiangolo.com/advanced/custom-request-and-route/

**Comment by Kludex** [comment]
Maybe it's worth saying that `BaseHTTPMiddleware` based middlewares should be avoided as well. :kissing:

**Comment by Dester507** [comment]
> Don't you need to pass the request on through? What if you mult. middlware? `call_next(request)` probably what you want. You should also be aware that processing the body in the middleware is something to try and avoid. Instead, create a custom request: https://fastapi.tiangolo.com/advanced/custom-request-and-route/

I need to call this APIRoute that i get from xml body. I have only method_name and attrs.

**Comment by Dester507** [comment]
> Don't you need to pass the request on through? What if you mult. middlware? `call_next(request)` probably what you want. You should also be aware that processing the body in the middleware is something to try and avoid. Instead, create a custom request: https://fastapi.tiangolo.com/advanced/custom-request-and-route/

For example, I send request  for this url(http://127.0.0.1:8000/xml/) with this body.
```xml
<?xml version="1.0" encoding="ASCII"?>
<methodCall>
    <methodName>method_name</methodName> # Method Name
    <params>
        <param>
            <value>
                <string> # Value Type
                    TEXT
                </string>
            </value>
        </param>

        <param>
            <value>
                <int> # Value Type
                    NUMBER
                </int>
            </value>
        </param>

        <param>
            <value>
                <nil/>
            </value>
        </param>
    </params>
</methodCall>
```
from this xml i parse method_name and attrs and next step is to call APIRoute for this func (where APIRoute.__dict__["name"] == method_name)

**Comment by Nathsad1996** [comment]
@Dester507  how do you find solution for this one ? I have the same problem

**Comment by Dester507** [comment]
Sorry i don't remember how, but i have git repo with solution.

**Comment by Nathsad1996** [comment]
Can you share it with me ?

**Comment by Dester507** [comment]
Send me your github profile, i'll give access for you

**Comment by Nathsad1996** [comment]
https://github.com/Nathsad1996

**Comment by Dester507** [comment]
Check in my repositories
Rep name -> prototype-API
After that you can find custom middleware in main.py
